### PR TITLE
Shortcut dat copy

### PR DESCRIPTION
**Pull request by dham**
Currently dat copying creates a par loop. This is slow and can dominate runtime on a small adjoint problem. In fact, numpy knows how to copy numpy arrays very quickly, so in the simplest case we should just do this.

This could possibly also be extended to the case where there is a subset.

### CODE REVIEW

File: pyop2/base.py [hunk 1]
```diff
@@ -1579,6 +1579,9 @@ def copy(self, other, subset=None):
         :arg subset: A :class:`Subset` of elements to copy (optional)"""
         if other is self:
             return
+        if subset is None:
+            other.data[:] = self.data
```
Comment (wence-) [inline pyop2/base.py:1583]: `self.data_ro`.

If you want to handle the subset case:
```
if subset is None:
    other.data[:] = self.data_ro
else:
    other.data[subset.indices] = self.data_ro[subset.indices]
```

Comment (wence-) [reply pyop2/base.py:1583]: I think you need to assert that the datasets for `other` and `self` match (and that if there's a subset that it also matches).

Comment (connorjward) [reply pyop2/base.py:1583]: If we do this then can we delete `_copy_parloop`?

**Comment by connorjward** [comment]
Should we do the same for `zero` as well?

**Comment by Melina97** [comment]
I ran into a type error so changed 'a += b' to np.add(a, b, out=a, casting='unsafe'), is that ok?

File: pyop2/base.py [hunk 2]
```diff
@@ -1698,19 +1653,6 @@ def _iop_kernel(self, op, globalp, other_is_self, dtype):
...
-        globalp = False
-        if np.isscalar(other):
-            other = _make_object('Global', 1, data=other)
-            globalp = True
-        elif other is not self:
-            self._check_shape(other)
-        args = [self(INC)]
-        if other is not self:
-            args.append(other(READ))
-        par_loop(self._iop_kernel(op, globalp, other is self, other.dtype), self.dataset.set, *args)
```
Comment (connorjward) [inline pyop2/base.py:1711]: Do we use `_iop_kernel` anywhere else? If not that can go.

File: pyop2/base.py [hunk 3]
```diff
@@ -1825,19 +1767,41 @@ def __truediv__(self, other):
 
     def __iadd__(self, other):
         """Pointwise addition of fields."""
-        return self._iop(other, operator.iadd)
+        from numbers import Number
+        if other is not None:
+            if isinstance(other, Number):
+                self.data[:] += other
+            else:
+                np.add(self.data[:], other.data_ro, out=self.data[:], casting="unsafe")
```
Comment (connorjward) [inline pyop2/base.py:1775]: I don't think we should be doing this as its behaviour is quite undefined. We should only add things if their shapes match. If you look in `_iop` above you'll see that we call `_check_shape`. Perhaps we should call that here.

Comment (Melina97) [reply pyop2/base.py:1775]: Yes, that sounds much better. Thanks!

**Comment by dham** [comment]
OK I have Firedrake tests running on this now. @Melina97 can you check whether `_iop` can be deleted as @connorjward suggested. If that works, we can merge this.

**Comment by Melina97** [comment]
I was prompted to do that last merge in order to push the changes, is that alright?

**Comment by dham** [comment]
> I was prompted to do that last merge in order to push the changes, is that alright?

That part is fine, but now that Connor has fixed the builder, we can see that there are actual test fails that need looking at.

**Comment by connorjward** [comment]
The tests that are failing suggest that being able to add things with different data types is actually what we want. I think setting `casting="unsafe"` but with `_check_shape` is probably the way to go.

File: pyop2/base.py [hunk 4]
```diff
@@ -1907,19 +1817,45 @@ def __truediv__(self, other):
...
-        return self._iop(other, operator.itruediv)
+        from numbers import Number
+        if isinstance(other, Number):
+            self.data[:] /= other
+        else:
+            self._check_shape(other)
+            np.true_divide(self.data[:], other.data_ro, out=self.data[:], casting="unsafe")
+        return self
 
     __idiv__ = __itruediv__  # Python 2 compatibility
```
Comment (connorjward) [inline pyop2/base.py:1860]: Can you delete this line at the same time? Python 2 is long gone.

**Review by connorjward — CHANGES_REQUESTED** [review]
Just 2 comments. Looks good otherwise.

File: pyop2/base.py [hunk 5]
```diff
@@ -1620,38 +1620,11 @@ def zero(self, subset=None):
...
-
-                knl = _make_object('Kernel', knl, 'zero')
-                self._zero_kernels[(self.dtype, self.cdim)] = knl
-            loop = _make_object('ParLoop', knl,
-                                iterset,
-                                self(WRITE))
-            loops[iterset] = loop
-        loop.compute()
+            self.data[subset.indices] = 0
+        return self
```
[the remark this answers is not in the document]
Comment (Melina97) [reply pyop2/base.py:1627]: Oh you're right, I'll remove that.

**Review by connorjward — APPROVED** [review]

**Comment by connorjward** [comment]
I think that this is ready to go. The [upstream checks](https://github.com/firedrakeproject/firedrake/pull/2158) are now passing.

The changes I made from Melina's work are:
- Reverted the changes to `__iadd__`, `__imul__` etc. They were breaking for mixed dats and aren't a current bottleneck like `zero` and `copy` are. We could address it in a separate PR at some point.
- Created and used `subset.owned_indices` instead of `subset.indices`. The reason being we were getting strange failures due to trying to index elements inside the dat's halo.

File: pyop2/base.py [hunk 6]
```diff
@@ -1620,38 +1627,12 @@ def zero(self, subset=None):
         """Zero the data associated with this :class:`Dat`
 
         :arg subset: A :class:`Subset` of entries to zero (optional)."""
-        if hasattr(self, "_zero_parloops"):
-            loops = self._zero_parloops
+        if subset is None:
+            self.data[:] = 0
```
Comment (wence-) [inline pyop2/base.py:1631]: There's an optimisation available here where you set `self._data[:] = 0` which fills in the halo region, and then set the `halo_valid` property (or whatever it is called) to `True`.

File: pyop2/base.py [hunk 7]
```diff
@@ -1661,28 +1642,12 @@ def copy(self, other, subset=None):
...
-            data = [loopy.GlobalArg("self", dtype=self.dtype, shape=(self.cdim,)),
-                    loopy.GlobalArg("other", dtype=other.dtype, shape=(other.cdim,))]
-            knl = loopy.make_function([domain], [insn], data, name="copy", target=loopy.CTarget(), lang_version=(2018, 2))
-
-            self._copy_kernel = _make_object('Kernel', knl, 'copy')
-        return _make_object('ParLoop', self._copy_kernel,
-                            subset or self.dataset.set,
-                            self(READ), other(WRITE))
+        if subset is None:
+            other.data[:] = self.data_ro
```
Comment (wence-) [inline pyop2/base.py:1646]: Similar kind of comment applies here. if `self.halo_valid` then copy the full array including halo regions (and set `other.halo_valid`).

File: pyop2/base.py (hunk 6 above)
Comment (connorjward) [reply pyop2/base.py]: Done.

File: pyop2/base.py (hunk 7 above)
Comment (connorjward) [reply pyop2/base.py]: Done.

**Pull request merged**